### ISSUE TITLE
Warnings and Errors

### ISSUE THREAD

**Issue by earthlng** [hultqvist/negotiator#30]
Hi, I tested Negotiator on FF55 nightly and got the following warnings and errors:

```
1491226601686	addons.webextension.negotiator@silentorbit.com	WARN	Loading extension 'negotiator@silentorbit.com': Reading manifest: Error processing background.persistent: Event pages are not currently supported. This will run as a persistent background page.
1491226601694	addons.webextension.negotiator@silentorbit.com	WARN	Loading extension 'negotiator@silentorbit.com': Reading manifest: Error processing minimum_edge_version: An unexpected property was found in the WebExtension manifest.
1491226601695	addons.webextension.negotiator@silentorbit.com	WARN	Loading extension 'negotiator@silentorbit.com': Reading manifest: Error processing update_url: An unexpected property was found in the WebExtension manifest.
1491226601695	addons.webextension.negotiator@silentorbit.com	WARN	Loading extension 'negotiator@silentorbit.com': Reading manifest: Error processing options_page: An unexpected property was found in the WebExtension manifest.
1491226601695	addons.webextension.negotiator@silentorbit.com	WARN	Loading extension 'negotiator@silentorbit.com': Reading manifest: Error processing inprivate: An unexpected property was found in the WebExtension manifest.
1491226601695	addons.webextension.negotiator@silentorbit.com	WARN	Loading extension 'negotiator@silentorbit.com': Reading manifest: Error processing -ms-preload: An unexpected property was found in the WebExtension manifest.
1491226601718	addons.webextension.negotiator@silentorbit.com	WARN	Please specify whether you want browser_style or not in your browser_action options.
```

background.html
`The character encoding of the HTML document was not declared. The document will render with garbled text in some browser configurations if the document contains characters from outside the US-ASCII range. The character encoding of the page must be declared in the document or in the transfer protocol.`

Most importantly when I tried to add a `New action` it reported an error
```
Error 83: /script/page-options.js
ReferenceError: event is not defined
```
... in errors.js:15:5

**Comment by earthlng** [comment]
created a [pull request](https://github.com/hultqvist/negotiator/pull/31) for the reported error

**Comment by earthlng** [comment]
A separate branch for Firefox with a custom manifest would be neat to get rid of those Warnings. You could also remove the Edge parts and maybe other things as well and build the Firefox extension from that branch.

**Comment by hultqvist** [comment]
Hi @earthlng , I have been busy. Once I get time I will look at all your contributions in one go.

**Comment by earthlng** [comment]
Hi, cool to see you are back :)
I also noticed that "Local only" seems kind of broken unfortunately, because most of the updating functions are not called. And instead of localStorage maybe you could use storage.local because ...

> Although this API is similar to Window.localStorage it is recommended that you don't use Window.localStorage in add-on code. Firefox will clear data stored by add-ons using the localStorage API in various scenarios where users clear their browsing history and data for privacy reasons, while data saved using the storage.local API will be correctly persisted in these scenarios.

[source](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/storage/local)

**Comment by hultqvist** [comment]
Today I've looked at your PR and accepted them all.
I also fixed a small bug an a major one which might have made it unusable in FF before the fix.
Right now I've tested it on stable FF and stable FF on Android.

The remaining warnings are from properties used in other browsers so those are there intentionally left there.

I hope this fix the errors.

**Comment by earthlng** [comment]
Thank you. I'll test the new version once it's available on AMO and will let you know if something is still broken.

**Comment by earthlng** [comment]
It still has problems. "error trying to access dead object" happened again. Local only is broken and the workaround of using Browser sync also no longer works because it now displays an error message "please enable sync" or something like that; maybe that's caused by a change in FF idk.

**Comment by hultqvist** [comment]
Hm, the dead object is a rather vaguely documented feature in FF. I haven't found any real explanation why it's a good idea considering alternatives.

Still I managed to fix some dead object issues last time, maybe there are other places that need the same fix. That is JSON.parse(JSON.stringify(object))

@earthlng can you open the debug console and see where exactly those errors occurred? File and line number.

As before, I can't say when I will have time to look at it, though I will analyze any pull request faster this time.

**Comment by earthlng** [comment]
Hi, it's me again, sorry xD

* localStorage problem: it gets deleted when someone clears cookies. And if cookies are blocked by default and only allowed on a per-site basis then localStorage doesn't work at all.

It's better to use `chrome.storage.local`. I would assign a new variable `var storage = chrome.storage.local` and use that instead. You would only have to rework sync.chrome.js and use the variable everywhere. When someone switches to "sync" you only have to update the storage variable: `storage = chrome.storage.sync`. This way you could get rid of all the localStorage stuff, and have less and cleaner code.
'local' should ideally be the default because 'storage.sync' is not available in Firefox ESR 52.x (https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/storage/sync#Browser_compatibility)

I would also consider getting rid of syncNegotiator because I doubt that anyone who would use your extension to remove Referrer headers for privacy reasons, etc, would want to send their configs to a remote server. It leaves kind of a foul taste that this option is even available tbh. Removing that would bring the code-size down even more.

* getBackgroundPage(): doesn't work in private/incognito windows unless Firefox starts in the Private-only mode. Otherwise 'b' is always null in private windows. see https://discourse.mozilla.org/t/getbackgroundpage-in-private-mode/14046.
IDK if you want to fix that at all but it would be nice if you did. IMHO your extension has the potential to become the de-facto fan-favorite to deal with anything Header-related.

To fix it you could implement the workaround that is mentioned in the link above, or what I personally would do is get rid of it completely: make the necessary functions self-contained, then load those function-files in the options pages directly, and keep local variables of the settings you need.
That approach would be a lot easier if you used a single options.html page and added "tabs" that show/hide the divs for each page. You would only have to keep the settings variable(s) synced between the background page and one options page, instead of 4 html pages.

* deleting an action that is assigned to 1 or more filters results in an error message:
`/script/page.js TypeError: b.actions[f.filter] is undefined`
  * a check that prevents deleting such an action would be nice

**Comment by hultqvist** [comment]
localStorage is now replaced with chrome.storage.local. Only in this repo at the moment, not enough for a release.

To defend NegotiatorSync my motivation was that the user could host their own server.

- The data wouldn't have to be synchronised with the google storage.
- It enables synchronisation across browsers(Chrome - Firefox).
- It's not limited by the quotas of chrome.storage.sync which currently is 512 items(I have way more filters) or 8kB per item(all filters wouldn't fit in one storage item.

That said the work on the hosted sync server has stalled and I'm not using it myself.
I do see the need in the future where I would like to manage my filters on a desktop and have them synchronised to my Firefox on android.
For now the code stays there.

About the rest, background page issues, single options page and blocking deletion of actions in use.
These are good ideas and would have to be solved all in one go.
I'll separate these in a new issue.
